### ISSUE TITLE
Pagy Extra is not working correctly with Meilisearch

### ISSUE THREAD

**Issue by doutatsu** [meilisearch/meilisearch-rails#263]
**Description**
I am using [Pagy Extra](https://ddnexus.github.io/pagy/docs/extras/meilisearch/) to support the pagination on Meilisearch. I am seeing an issue that items count is not being respected on some pages, but not the others

**Expected behavior**
I set number of results per page at 24 items. If there is a next page available, I expect the current page to return exactly 24 items. Only the last page can have less than 24
```rb
## Controller      
pagy_meilisearch(search_items, items: 24, size: [1, 3, 3, 1])

## Service
def search_items
      Series.pagy_search(
        params[:search_term],
        highlight_pre_tag: '<span class="highlight">',
        highlight_post_tag: '</span>',
        matching_strategy: 'all',
        sort: sort,
        filter: filters
      )
end
```

**Current behavior**
Sometimes pagy returns 23 items instead of 24, even though more pages are available

**Screenshots or Logs**
This is the returned data for incorrect number of items. As can be seen, pagy items is set to 24, but it returns only 23 results, even when there are multiple pages to go
<img width="175" alt="image" src="https://github.com/meilisearch/meilisearch-rails/assets/4270980/78466f52-f472-409e-b975-f70797b62149">

Here is another page, this time correctly returns 24 items:
<img width="140" alt="image" src="https://github.com/meilisearch/meilisearch-rails/assets/4270980/6733f4fd-0d6b-4db6-a653-93ea18848a0d">

**Environment (please complete the following information):**
- OS: Locally (MacOS) and in Production (Linux)
- Meilisearch server version: 
- meilisearch-rails version: 
- Rails version:

**Comment by doutatsu** [comment]
An update -> I managed to fix it by clearing the index and then reindexing everything from scratch. This feels like a bug for sure to me

**Comment by brunoocasali** [comment]
Interesting to know, and thanks for your debugging. I want the inputs from @oluademola because that could probably be used by the engine team.

**Comment by Menelle** [comment]
We are experiencing the exact same bug. We attempted to reindex the catalog multiple times now but it didn't fix the issue unfortunately.

**Comment by Menelle** [comment]
Update. Not sure if related. In our case, we found out at that the issue was caused by our pundit scope daisy chain.
Meilisearch returned results matching the search as expected,  then the AR scope from pundit gem is applied, then the pagination iss generated with pagy onto those results, creating those random holes on each page.
Possible solution is to pass the pundit (or metal) scoped ids as meilisearch filter manually to retrieve only required objects, but that's an extra db pluck call. In case of simple scoping condition (not based on user), multiple indexes seems cleaner.

**Comment by doutatsu** [comment]
Had this issue again - it's consistent the moment you delete an entry, pagination goes whack. And you have to clear, and then index again to make it work again. Should be easy to reproduce

**Comment by brunoocasali** [comment]
Hi @doutatsu, can you make a cURL request with the same arguments when you experience that faulty moment? I want to isolate the issue from an integration/engine perspective.

**Comment by doutatsu** [comment]
Here is more context. My call is as follows in the service:

```rb
      MangaSeries.pagy_search(
        params[:search_term],
        highlight_pre_tag: '<span class="highlight">',
        highlight_post_tag: '</span>',
        matching_strategy: 'all',
        sort:,
        filter: filters
      )
```

these are the params used, after reproduction:

```rb
{
  nil, 
  :highlight_pre_tag=>"<span class=\"highlight\">", 
  :highlight_post_tag=>"</span>", 
  :matching_strategy=>"all", 
  :sort=>["created_at_unix:desc"], 
  :filter=>["has_active_sources = TRUE"]
}
```

In the controller, I am doing this:

```rb
  pagy_meilisearch(
    MangaSearch::FindSeries.call(params),
    items: 100,
    size: [1, 2, 2, 1]
  )
```

Now, I've done the same call using cURL as suggested by @brunoocasali. And I can see the issue now - it does return 6 per page, **but** it returns documents that were deleted! Even though a particular Series has been deleted from the database, `reindex!` doesn't remove it. So cURL returns the removed document, while at the controller level I must be filtering out `nil` somewhere, as I serialize the found series.

So seems like the issue is not with pagination, but rather `reindex!` not removing removed series in my case:
<img width="1390" alt="Screenshot 2023-06-19 at 10 32 45" src="https://github.com/meilisearch/meilisearch-rails/assets/4270980/c41b60c6-1965-439e-9d01-c4f2faea48ff">
<img width="522" alt="Screenshot 2023-06-19 at 10 32 34" src="https://github.com/meilisearch/meilisearch-rails/assets/4270980/8164cbdb-f194-4113-8681-ed450a49904d">

After going ahead and manually using `index.delete_document` "pagination" is back to normal

**Comment by doutatsu** [comment]
Unless I misunderstand the use of re-indexing? I suppose I should be able to fix this by making sure to call `index.delete_document` when series has been deleted, but I was hoping that `reindex!` would do the same - aka delete documents that exist in index, but not in the table, while updating data for those existing

**Comment by oluademola** [comment]
Hello @doutatsu. I think there might be a misunderstanding. Deleting a row in your table does not effectively delete it in the meilisearch index with "reindex!". The "reindex!" method does not delete the documents in the meilisearch index. Instead, it adds and then replaces documents with the same ID, while the rest of the documents remain in the index. It reindexes your models into meilisearch using this [route](https://www.meilisearch.com/docs/reference/api/documents#add-or-replace-documents). Does this answer your question?

**Comment by doutatsu** [comment]
@oluademola I see - what is the reason for that? When I think `reindex!`, I assumed it would completely update the index, meaning updating documents, deleting removed documents and adding new documents. Making it a nice, easy-to-use method to make sure the index is identical to the current state of the table.

**Comment by brunoocasali** [comment]
Yep, @oluademola is right @doutatsu, and you are also suitable to think that way. This should be improved by making the `reindex!` create a new temporary index add the data, and then switch it to the current index name. So we can run it in production without downtime.

I've created a new issue #267 to track that improvement. But I assume this is not a bug anymore @doutatsu right?
